compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -474,6 +483,179 @@ pub fn normalize_param_env_or_error<'tcx>(
     ty::ParamEnv::new(tcx.mk_clauses(&predicates), unnormalized_env.reveal())
 }
 
+#[derive(Debug)]
+pub enum EvaluateConstErr {
+    HasGenericsOrInfers,
+    InvalidConstParamTy(ErrorGuaranteed),
+    CTFEFailure(ErrorGuaranteed),
```

Comment:
I commented below, but `CTFEFailure` is a bit of a mouthful. Not totally sure what a shorter name is, but at least could we document what kinds of failures it's meant to represent?

Reply:
This also perhaps could just be called `ErrorReported` since it's always downstream of an error being reported. Also, we may be able to just fold `InvalidConstParamTy` into it. If that's not worthwhile to do this PR, could you leave a FIXME?

Reply:
I'm not entirely sure if we can fold them into the same variant or not. I would be a bit nervous about normalizing away some ill formed constant to a `ConstKind::Error` with a `bug!` and then never actually emitting an error about its type not matching up with the parameter it was an argument to.